Makefile.COMMON

```diff
@@ -0,0 +1,103 @@
+# Copyright 2015 The Prometheus Authors
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This file provides common Makefile infrastructure for several Prometheus
+# components. This includes make tasks for downloading Go, setting up a
+# self-contained build environment, fetching Go dependencies, building
+# binaries, running tests, and doing release management. This file is intended
+# to be included from a project's Makefile, which needs to define the following
+# variables, at a minimum:
+#
+# * VERSION - The current version of the project in question.
+# * TARGET  - The desired name of the built binary.
+#
+# Many of the variables defined below are defined conditionally (using '?'),
+# which allows the project's main Makefile to override any of these settings, if
+# needed. See also:
+#
+# https://www.gnu.org/software/make/manual/html_node/Flavors.html#Flavors.
+#
+# The including Makefile may define any number of extra targets that are
+# specific to that project.
+
+VERSION ?= $(error VERSION not set in including Makefile)
+TARGET  ?= $(error TARGET not set in including Makefile)
+
+SRC    := $(shell find . -type f -name "*.go" ! -path "./.build/*")
+GOOS   ?= $(subst Darwin,darwin,$(subst Linux,linux,$(subst FreeBSD,freebsd,$(shell uname))))
+GOARCH ?= $(subst x86_64,amd64,$(patsubst i%86,386,$(shell uname -m)))
+
+ifeq ($(GOOS),darwin)
+	RELEASE_SUFFIX ?= -osx$(shell sw_vers -productVersion)
+endif
+
+GO_VERSION ?= 1.4.1
+GOURL      ?= https://golang.org/dl
+GOPKG      ?= go$(GO_VERSION).$(GOOS)-$(GOARCH)$(RELEASE_SUFFIX).tar.gz
+GOROOT     ?= $(CURDIR)/.build/go
```

Comment:
I think GOPKG/GOROOT shouldn't pick up things from the environment, as that may pick up random versions of packages and local changes by the user. This prevents a hermetic build.